test/hexpm/accounts/users_test.exs

```diff
@@ -0,0 +1,237 @@
+defmodule Hexpm.Accounts.UsersTest do
+  use Hexpm.DataCase, async: true
+
+  alias Hexpm.Accounts.Users
+
+  describe "update_profile/3 when user belongs to an organization" do
+    test "updates full_name" do
+      organization = insert(:organization, user: build(:user, full_name: "Old Full Name"))
+
+      {:ok, updated_user} =
+        Users.update_profile(
+          organization.user,
+          %{"full_name" => "New Full Name"},
+          audit: {build(:user), "UA"}
+        )
+
+      assert %{full_name: "New Full Name"} = updated_user
+    end
+
+    test "updates handles" do
+      organization = insert(:organization, user: build(:user))
+
+      {:ok, updated_user} =
+        Users.update_profile(
+          organization.user,
+          %{
+            "handles" => %{
+              "twitter" => "twitter",
+              "github" => "github",
+              "elixirforum" => "elixirforum",
+              "freenode" => "freenode",
+              "slack" => "slack"
+            }
+          },
+          audit: {build(:user), "UA"}
+        )
+
+      assert %{
+               twitter: "twitter",
+               github: "github",
+               elixirforum: "elixirforum",
+               freenode: "freenode",
+               slack: "slack"
+             } = updated_user.handles
+    end
+
+    test "audits user.update action" do
+      organization =
+        insert(:organization, user: build(:user, username: "organization.user", emails: []))
+
+      current_user = insert(:user)
+
+      {:ok, _} =
+        Users.update_profile(
+          organization.user,
+          %{},
+          audit: {current_user, "UA"}
+        )
+
+      # TODO: attach this audit_log to organization as well?
+      assert [%{action: "user.update", params: %{"username" => "organization.user"}}] =
+               Hexpm.Accounts.AuditLogs.all_by(current_user)
+    end
+
+    test "inserts public email if it doesn't exist yet" do
+      organization = insert(:organization, user: build(:user, emails: []))
+      current_user = insert(:user)
+
+      {:ok, _updated_user} =
+        Users.update_profile(
+          organization.user,
+          %{"public_email" => "public@example.com"},
+          audit: {current_user, "UA"}
+        )
+
+      email = Users.get_email("public@example.com")
+      assert email.user_id == organization.user.id
+      assert email.public == true
+
+      # TODO: these audit_logs shall belong to organization or organization user?
+      assert [%{action: "email.public"}, %{action: "email.add"}, _user_update] =
```

Comment:
the `audit_log.user` is current_user, which may not be accurate, do we need to modify audit_data before calling `audit`?

Reply:
The user is the person who performed the action, which cannot be an organization, so this looks correct.

Reply:
Okay, the user should be the current user.

But the problem is that we can't know which organization this `audit_log` is for, because
1. we are not passing any info related to organization to `audit/4`
2. only `[:email, :primary, :public, :primary, :gravatar]` are saved when serializing `email`, so we cannot fetch organization by calling `audit_log.params.user_id`, then get organization from this user.

So I think we need to pass organization info to `audit/4`, either by:
1. passing `{organization, conn.assigns.user_agent}` as audit_data in `OrganizationController.update_profile/2`
   - the implementation is easy
   - but the `audit_log.user_id` would be `nil` (we don't know who performed the action anymore)
2. passing `Map.merge(email, %{organization: %{id: ...}})` as params to `audit/4`
   - we keep the `audit_log.user_id` as current_user
   - but the implementation is not easy (fetching user.organization in `Users` module and merging params when calling `audit/4`)

Let me know which option is better, thank you!

Reply:
Or option3: pass `{user, org, UA}` to `audit/4` so we can record action performer and organization at the same time?

Reply:
We can call it as: `AuditLog.build(current_user, user_agent, "email.*", {organization, email})`. We unfortunately have fetch user.organization in `Users` but we don't have merge params.

We have to change `extract_params("email.*")` to return the organization so that we can add it to `AuditLog.organization_id` here: https://github.com/hexpm/hexpm/blob/master/lib/hexpm/accounts/audit_log.ex#L45.

Does that make sense?

Reply:
Yes, of course.

So I will keep the old `extract_params` functions for `email.*`, and add new ones for `{organization, email}`

Reply:
Updated in https://github.com/hexpm/hexpm/pull/878/commits/da0e28f02d809c1109c8118857717f4fad77beae